src/base.c

```diff
@@ -185,24 +194,57 @@ base_extent_bump_alloc_helper(edata_t *edata, size_t *gap_size, size_t size,
 	assert(edata_bsize_get(edata) >= *gap_size + size);
 	edata_binit(edata, (void *)((byte_t *)edata_addr_get(edata) +
 	    *gap_size + size), edata_bsize_get(edata) - *gap_size - size,
-	    edata_sn_get(edata));
+	    edata_sn_get(edata), base_edata_is_reused(edata));
 	return ret;
 }
 
 static void
-base_extent_bump_alloc_post(base_t *base, edata_t *edata, size_t gap_size,
-    void *addr, size_t size) {
+base_edata_heap_insert(tsdn_t *tsdn, base_t *base, edata_t *edata) {
+	malloc_mutex_assert_owner(tsdn, &base->mtx);
+
+	size_t bsize = edata_bsize_get(edata);
+	assert(bsize > 0);
+	/*
+	 * Compute the index for the largest size class that does not exceed
+	 * extent's size.
+	 */
+	szind_t index_floor = sz_size2index(bsize + 1) - 1;
+	edata_heap_insert(&base->avail[index_floor], edata);
+}
+
+/*
+ * Only can be called by top-level functions, since it may call base_alloc
+ * internally when cache is empty.
+ */
+static edata_t *
+base_alloc_base_edata(tsdn_t *tsdn, base_t *base) {
+	edata_t *edata;
+
+	malloc_mutex_lock(tsdn, &base->mtx);
+	edata = edata_avail_first(&base->edata_avail);
+	if (edata != NULL) {
+		edata_avail_remove(&base->edata_avail, edata);
+	}
+	malloc_mutex_unlock(tsdn, &base->mtx);
+
+	if (edata == NULL) {
+		edata = base_alloc_edata(tsdn, base);
+	}
+
+	return edata;
+}
+
+static void
+base_extent_bump_alloc_post(tsdn_t *tsdn, base_t *base, edata_t *edata,
+    size_t gap_size, void *addr, size_t size) {
 	if (edata_bsize_get(edata) > 0) {
-		/*
-		 * Compute the index for the largest size class that does not
-		 * exceed extent's size.
-		 */
-		szind_t index_floor =
-		    sz_size2index(edata_bsize_get(edata) + 1) - 1;
-		edata_heap_insert(&base->avail[index_floor], edata);
+		base_edata_heap_insert(tsdn, base, edata);
+	} else {
+		/* Freed base edata_t stored in edata_avail. */
+		edata_avail_insert(&base->edata_avail, edata);
 	}
 
-	if (config_stats) {
+	if (config_stats && !base_edata_is_reused(edata)) {
```

Comment:
This seems to be the main reason why we are maintaining another `edata` for the allocated stack. Why not directly use the guarded bit of the `edata` that already exists?

Reply:
The edata already exists is from the initial base_block, and the allocated stack is bump allocated off that block. So there's no way to merge it back into the block because of the bump alloc -- that's why we need a new and separate edata_t to manage the returned tcache_stack (it gets `base_edata_heap_insert` into the available extents for future allocations).

In other words, the base_block bump allocates the initial tcache_stack, and when deallocating, the tcache_stack and the base_block are not guaranteed to be continuous, thus requiring to be managed separately with a new edata_t.

As for the guarded bit, yes the only reason that we use that to mark and track "reused", is to avoid double counting stats.

Reply:
I see. So the life cycle of this new edata is:
1. It is newed/retrieved from `edata_avail` and initialized when a tcache_stack is allocated;
2. When the stack is deallocated, representing the allocation, this edata is put into `avail` for the memory to be reused;
3. If the stack is reused up, the edata is inserted back to `edata_avail` for it to be reused.

Is my understanding correct?

Another quick question, can you remind me why we did not need an `else` above to handle `edata_bsize_get(edata) == 0` case before?

Reply:
> Is my understanding correct?

Yes!

> why we did not need an else above to handle edata_bsize_get(edata) == 0 case before

Previously, all edatas were part of the base_block_t: https://github.com/jemalloc/jemalloc/blob/a442d9b895935ac872e7ccc705213537bc747c19/include/jemalloc/internal/base.h#L40

Since the # of base blocks were limited (similar to grow_retained: https://github.com/jemalloc/jemalloc/blob/a442d9b895935ac872e7ccc705213537bc747c19/include/jemalloc/internal/base.h#L65), we didn't bother reusing them. Instead every time we allocate a new block which contains a edata_t inside. Technically we could still reuse the edata_t once a block is drained, but the savings would be more like a few hundred bytes which is a bit hard to justify the complexity.

Reply:
I see. Then I assume the original edata will also be reused with this new `else` added?

Reply:
That's right. So we are also getting a tiny bit of savings from that, with this PR.